proof/refine/AARCH64/TcbAcc_R.thy

```diff
@@ -1832,56 +1994,700 @@ crunch typ_at'[wp]: removeFromBitmap "\<lambda>s. P (typ_at' T p s)"
+  apply (frule tcbQueueHead_iff_tcbQueueEnd)
+  by (clarsimp simp: tcbQueueEmpty_def list_queue_relation_def queue_end_valid_def)
+
+lemma obj_at'_tcbQueueEnd_ksReadyQueues:
+  "\<lbrakk>list_queue_relation ts queue nexts prevs;
+    \<forall>t. (inQ d p |< tcbs_of' s') t \<longleftrightarrow> t \<in> set ts;
+    pspace_aligned' s'; pspace_distinct' s'\<rbrakk>
+   \<Longrightarrow> \<not> tcbQueueEmpty queue \<longrightarrow> obj_at' (inQ d p) (the (tcbQueueEnd queue)) s'"
+  by (fastforce dest!: tcbQueueEnd_ksReadyQueues intro: aligned'_distinct'_ko_wp_at'I
+                 simp: obj_at'_real_def opt_map_def opt_pred_def split: option.splits)
+
+lemma thread_get_exs_valid[wp]:
+  "tcb_at tcb_ptr s \<Longrightarrow> \<lbrace>(=) s\<rbrace> thread_get f tcb_ptr \<exists>\<lbrace>\<lambda>_. (=) s\<rbrace>"
+  by (clarsimp simp: thread_get_def get_tcb_def gets_the_def gets_def return_def get_def
+                     exs_valid_def tcb_at_def bind_def)
+
+lemma ethread_get_sp:
+  "\<lbrace>P\<rbrace> ethread_get f ptr
+   \<lbrace>\<lambda>rv. etcb_at (\<lambda>tcb. f tcb = rv) ptr and P\<rbrace>"
+  apply wpsimp
+  apply (clarsimp simp: etcb_at_def split: option.splits)
+  done
+
+lemma ethread_get_exs_valid[wp]:
+  "\<lbrakk>tcb_at tcb_ptr s; valid_etcbs s\<rbrakk> \<Longrightarrow> \<lbrace>(=) s\<rbrace> ethread_get f tcb_ptr \<exists>\<lbrace>\<lambda>_. (=) s\<rbrace>"
+  apply (frule (1) tcb_at_is_etcb_at)
+  apply (clarsimp simp: ethread_get_def get_etcb_def gets_the_def gets_def return_def get_def
+                        is_etcb_at_def exs_valid_def bind_def)
+  done
+
+lemma no_fail_ethread_get[wp]:
+  "no_fail (tcb_at tcb_ptr and valid_etcbs) (ethread_get f tcb_ptr)"
+  unfolding ethread_get_def
+  apply wpsimp
+  apply (frule (1) tcb_at_is_etcb_at)
+  apply (clarsimp simp: is_etcb_at_def get_etcb_def)
+  done
+
+lemma threadGet_sp:
+  "\<lbrace>P\<rbrace> threadGet f ptr \<lbrace>\<lambda>rv s. \<exists>tcb :: tcb. ko_at' tcb ptr s \<and> f tcb = rv \<and> P s\<rbrace>"
+  unfolding threadGet_def setObject_def
+  apply (wpsimp wp: getObject_tcb_wp)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma in_set_ready_queues_inQ_eq:
+  "ready_queues_relation s s' \<Longrightarrow> t \<in> set (ready_queues s d p) \<longleftrightarrow> (inQ d p |< tcbs_of' s') t"
+  by (clarsimp simp: ready_queues_relation_def Let_def)
+
+lemma in_ready_q_tcbQueued_eq:
+  "ready_queues_relation s s'
+   \<Longrightarrow> (\<exists>d p. t \<in> set (ready_queues s d p))  \<longleftrightarrow> (tcbQueued |< tcbs_of' s') t"
+  apply (intro iffI)
+   apply clarsimp
+   apply (frule in_set_ready_queues_inQ_eq)
+   apply (fastforce simp: inQ_def opt_map_def opt_pred_def split: option.splits)
+  apply (fastforce simp: ready_queues_relation_def Let_def inQ_def opt_pred_def
+                  split: option.splits)
+  done
+
 lemma tcbSchedEnqueue_corres:
```

Comment:
the proof of this lemma looks rather similar to tcbSchedAppend_corres, I'm wondering if we know why this one is here and the other in Schedule_R

Reply:
I think I put the `corres` proofs where the old versions were. I think because the enqueue is used in more functions in the kernel than the append is, we do a lot of the enqueue stuff first. 

The proofs are quite similar, but unfortunately because the Haskell manipulates the next and prev pointers, they're not instances of a more general theorem.

Reply:
Yes, their placement was before your time, but you have the freshest view of them. Do you think there is any benefit to migrating them to be closer together (copy-pasta nearby is easier to notice and update in tandem), or would it make a bigger mess?

Reply:
I think the only benefit would be to make it easier to update both of them if they both need trivial changes. I'd rather not move them as part of this PR, if I do move them. Would maybe a comment saying `This proof shares many similarities with the proof of @{thm tcbSchedEnqueue_corres}` be a good compromise for now? 

If everyone else thinks I should move them, then I can do this in another small PR after this is merged (though I'd rather not accumulate half a dozen PRs, seeing that they all need review and nothing is arch split so it takes a long time to do anything)

Reply:
Yeah, maybe throw in a comment for that and the Dequeue equivalent.

Reply:
I've added the comment I wrote above just before `tcbSchedAppend_corres`